docs(framework) Update how to use strategies guide (#4522)
adap/flower: doc/source/how-to-use-strategies.rst

Use an existing strategy
------------------------

Flower comes with a number of popular federated learning Strategies which can be
instantiated as follows:

.. code-block:: python

    from flwr.common import Context
    from flwr.server.strategy import FedAvg
    from flwr.server import ServerApp, ServerAppComponents, ServerConfig

    def server_fn(context: Context):
        # Optional context-based parameters specification
        num_rounds = context.run_config["num-server-rounds"]
        config = ServerConfig(num_rounds=num_rounds)

        strategy = FedAvg(
            fraction_fit=context.run_config["fraction-fit"],
            fraction_evaluate=1.0,
        )
        return ServerAppComponents(strategy=strategy, config=config)

    # Create ServerApp
    app = ServerApp(server_fn=server_fn)

> A Strategy needs to be instantiated inside a function that takes ``Context`` and returns
> ``ServerAppComponents`` (e.g. named ``server_fn`` in this example). The parameters of

Reviewer: ```suggestion
A Strategy needs to be instantiated inside a function (e.g. ``server_fn`` in this example) that takes ``Context`` and returns a
``ServerAppComponents`` object. The parameters of
```

> the strategy can be customized by accessing them from the context or or hard-coding
> directly in the function. The ``server_fn`` function needs to be passed to a

Reviewer: ```suggestion
the strategy can be customized by accessing them from the context, which in turn
are set under the ``[tool.flwr.app.config]`` table in your ``pyproject.toml``. The ``server_fn`` function needs to be passed to a
```

Use an existing strategy
------------------------

Flower comes with a number of popular federated learning Strategies which can be
instantiated as follows:

.. code-block:: python

    from flwr.common import Context
    from flwr.server.strategy import FedAvg
    from flwr.server import ServerApp, ServerAppComponents, ServerConfig

    def server_fn(context: Context):
        # Optional context-based parameters specification
        num_rounds = context.run_config["num-server-rounds"]
        config = ServerConfig(num_rounds=num_rounds)

> 
>         strategy = FedAvg(

Reviewer 2: ```suggestion

        # Instantiate FedAvg strategy
        strategy = FedAvg(
```

            fraction_fit=context.run_config["fraction-fit"],
            fraction_evaluate=1.0,

>         )
>         return ServerAppComponents(strategy=strategy, config=config)

Reviewer 2: ```suggestion
        )

        return ServerAppComponents(strategy=strategy, config=config)
```

instantiated as follows:

.. code-block:: python

    from flwr.common import Context
    from flwr.server.strategy import FedAvg
    from flwr.server import ServerApp, ServerAppComponents, ServerConfig

    def server_fn(context: Context):
        # Optional context-based parameters specification
        num_rounds = context.run_config["num-server-rounds"]
        config = ServerConfig(num_rounds=num_rounds)

        # Instantiate FedAvg strategy
        strategy = FedAvg(
            fraction_fit=context.run_config["fraction-fit"],
            fraction_evaluate=1.0,
        )

        return ServerAppComponents(strategy=strategy, config=config)

    # Create ServerApp
    app = ServerApp(server_fn=server_fn)

A Strategy needs to be instantiated inside a function (e.g. ``server_fn`` in this
example) that takes ``Context`` and returns a ``ServerAppComponents`` object. The
parameters of the strategy can be customized by accessing them from the ``Context``
object, which in turn contains the values specified in the ``[tool.flwr.app.config]``
table in your ``pyproject.toml`` (a snippet is shown below).

.. code-block:: python

    ...
    [tool.flwr.app.config]

>     num - server - rounds = 10
>     fraction - fit = 0.5

Reviewer 2: ```suggestion
    num-server-rounds = 10
    fraction-fit = 0.5
```

        return fit_config

    def server_fn(context: Context):
        # Optional context-based parameters specification
        num_rounds = context.run_config["num-server-rounds"]
        config = ServerConfig(num_rounds=num_rounds)

        strategy = FedAvg(
            fraction_fit=context.run_config["fraction-fit"],
            fraction_evaluate=1.0,
            on_fit_config_fn=get_on_fit_config_fn(),
        )
        return ServerAppComponents(strategy=strategy, config=config)

    # Create ServerApp
    app = ServerApp(server_fn=server_fn)

The ``on_fit_config_fn`` can be used to pass arbitrary configuration values from server
to client and potentially change these values each round, for example, to adjust the
learning rate. The client will receive the dictionary returned by the
``on_fit_config_fn`` in its own ``client.fit()`` function. And while the values can be

> also passed directly form via the context this function can be a place of adding some
> randomness that can't be achieved by the context, which is static.

Reviewer: ```suggestion
also passed directly via the context this function can be a place to implement
finer control over the `fit` behaviour that may not be achieved by the context, which sets fixed values.
```
